### PR TITLE
Sip 185 debt shares

### PR DESCRIPTION
**Pull request by dbeal-eth**
Implements [SIP-185](https://sips.synthetix.io/sips/sip-185/)

Currently missing pieces:
* `SynthetixBridgeToOptimism` and `SynthetixBridgeToBase` changes to support actual migration of the shares
* Script to generate the `importAddrseses` transaction for migration time
* unit tests for `SynthetixDebtShare` itself

notes:
* Upon deployment, for both networks, this SIP will require running a one time`importAddresses` script to migrate the debt from the old `SynthetixState` to `SynthetixDebtShares`. how will this be done with the multisig on mainnet?
* SIP says "The precision of the debt shares will be 18 decimal places." In accordance with this, tests have been modified to accept higher tolerance for debt calculations.
* many of the tests in the `Liquidations` unit test should probably be moved to `Issuer` becuase the functionality being tested is actually in the `Issuer` contract. For now I have just adjusted the tests.

### CODE REVIEW

File: contracts/SynthetixDebtShare.sol [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+        uint balance = balanceOf(account);
+
+        if (balance == 0) {
+            return 0;
+        }
+
+        return balance.divideDecimal(totalSupply());
+    }
+
+    function sharePercentOnPeriod(address account, uint periodId) public view returns (uint) {
```
Comment (barrasso) [inline contracts/SynthetixDebtShare.sol:91]: [np] `sharePercentOnPeriod` could also be declared `external`

File: contracts/SynthetixDebtShare.sol [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+        emit Transfer(address(from), address(to), amount);
+    }
+
+    function importAddresses(address[] memory accounts, uint256[] memory amounts) public onlyOwner onlySetup {
+        for (uint i = 0; i < accounts.length; i++) {
+            mintShare(accounts[i], amounts[i]);
+        }
+    }
+
+    function finishSetup() public onlyOwner {
```
Comment (barrasso) [inline contracts/SynthetixDebtShare.sol:171]: [np] `importAddresses` and `finishSetup` can be declared `external`

**Review by barrasso — CHANGES_REQUESTED** [review]
Looks good overall -- just had some minor comments. Also, needs some moar coverage

File: contracts/FeePool.sol [hunk 3]
```diff
@@ -363,18 +330,14 @@ contract FeePool is Owned, Proxyable, LimitedSetup, MixinSystemSettings, IFeePoo
     function importFeePeriod(
         uint feePeriodIndex,
         uint feePeriodId,
-        uint startingDebtIndex,
         uint startTime,
         uint feesToDistribute,
         uint feesClaimed,
         uint rewardsToDistribute,
         uint rewardsClaimed
     ) public optionalProxy_onlyOwner onlyDuringSetup {
```
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply contracts/FeePool.sol:338]: wouldn't that be out of scope for this PR?

File: contracts/SynthetixDebtShare.sol (hunk 2 above)
Comment (dbeal-eth) [reply contracts/SynthetixDebtShare.sol]: https://github.com/Synthetixio/synthetix/pull/1601/commits/834ec9d650cf2905bed85769d99f8a5a30703d33

File: contracts/SynthetixDebtShare.sol (hunk 1 above)
Comment (dbeal-eth) [reply contracts/SynthetixDebtShare.sol]: https://github.com/Synthetixio/synthetix/pull/1601/commits/834ec9d650cf2905bed85769d99f8a5a30703d33

File: contracts/SynthetixDebtShare.sol [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+                return balances[account][uint(i)].amount;
+            }
+        }
+    }
+
+    function totalSupply() public view returns (uint) {
+        return totalSupplyOnPeriod[currentPeriodId];
+    }
+
+    function sharePercent(address account) public view returns (uint) {
```
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply contracts/SynthetixDebtShare.sol]: https://github.com/Synthetixio/synthetix/pull/1601/commits/834ec9d650cf2905bed85769d99f8a5a30703d33

File: contracts/FeePool.sol [hunk 5]
```diff
@@ -608,22 +570,16 @@ contract FeePool is Owned, Proxyable, LimitedSetup, MixinSystemSettings, IFeePoo
...
 
             // We can skip the period, as no debt minted during period (next period's startingDebtIndex is still 0)
-            if (nextPeriodStartingDebtIndex > 0 && lastFeeWithdrawal < _recentFeePeriodsStorage(i).feePeriodId) {
-                // We calculate a feePeriod's closingDebtIndex by looking at the next feePeriod's startingDebtIndex
-                // we can use the most recent issuanceData[0] for the current feePeriod
-                // else find the applicableIssuanceData for the feePeriod based on the StartingDebtIndex of the period
-                uint closingDebtIndex = uint256(nextPeriodStartingDebtIndex).sub(1);
+            if (lastFeeWithdrawal < _recentFeePeriodsStorage(i).feePeriodId) {
 
                 // Gas optimisation - to reuse debtEntryIndex if found new applicable one
```
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply contracts/FeePool.sol]: https://github.com/Synthetixio/synthetix/pull/1601/commits/bff2fb6400d807b30f10d4164883e2d641fc000c

File: contracts/FeePool.sol [hunk 6]
```diff
@@ -608,22 +570,16 @@ contract FeePool is Owned, Proxyable, LimitedSetup, MixinSystemSettings, IFeePoo
         // Go through our fee periods from the oldest feePeriod[FEE_PERIOD_LENGTH - 1] and figure out what we owe them.
         // Condition checks for periods > 0
         for (uint i = FEE_PERIOD_LENGTH - 1; i > 0; i--) {
-            uint next = i - 1;
-            uint nextPeriodStartingDebtIndex = _recentFeePeriodsStorage(next).startingDebtIndex;
 
             // We can skip the period, as no debt minted during period (next period's startingDebtIndex is still 0)
```
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply contracts/FeePool.sol]: https://github.com/Synthetixio/synthetix/pull/1601/commits/9043de18ee9abe340fddad027f8a782e2fc1676e

File: contracts/FeePool.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply contracts/FeePool.sol]: https://github.com/Synthetixio/synthetix/pull/1601/commits/9043de18ee9abe340fddad027f8a782e2fc1676e

File: contracts/BaseSynthetix.sol [hunk 7]
```diff
@@ -147,14 +140,13 @@ contract BaseSynthetix is IERC20, ExternStateToken, MixinResolver, ISynthetix {
     }
 
     function _canTransfer(address account, uint value) internal view returns (bool) {
-        (uint initialDebtOwnership, ) = synthetixState().issuanceData(account);
-
-        if (initialDebtOwnership > 0) {
+        if (issuer().debtBalanceOf(account, "sUSD") > 0) {
```
Comment (jacko125) [inline contracts/BaseSynthetix.sol:143]: Why not check whether the account has any debt shares and if 0, you could save more gas and skip this check for transferring ?

File: contracts/FeePool.sol [hunk 8]
```diff
@@ -580,15 +544,14 @@ contract FeePool is Owned, Proxyable, LimitedSetup, MixinSystemSettings, IFeePoo
...
+        ISynthetixDebtShare _debtShare = synthetixDebtShare();
 
-        (userOwnershipPercentage, debtEntryIndex) = _feePoolState.getAccountsDebtEntry(account, 0);
+        userOwnershipPercentage = _debtShare.sharePercent(account);
 
         // If they don't have any debt ownership and they never minted, they don't have any fees.
         // User ownership can reduce to 0 if user burns all synths,
         // however they could have fees applicable for periods they had minted in before so we check debtEntryIndex.
-        if (debtEntryIndex == 0 && userOwnershipPercentage == 0) {
+        if (userOwnershipPercentage == 0) {
```
Comment (jacko125) [inline contracts/FeePool.sol:554]: @dbeal-eth  Have to be careful here that they can still claim previous fee periods even if they now have burned all their debt / no current debt shares %. 

This is why it checked that their `debtEntryIndex == 0` as well which meant they never minted debt before.

File: contracts/BaseSynthetix.sol (hunk 7 above)
Comment (dbeal-eth) [reply contracts/BaseSynthetix.sol:143]: I think I just wanted to reuse the `issuer` dependency. You are right, checking the debt shares contract directly would make more sense.

File: contracts/FeePool.sol (hunk 8 above)
Comment (dbeal-eth) [reply contracts/FeePool.sol:554]: is there a unit test, or an easy way to write one?

Comment (jacko125) [reply contracts/FeePool.sol:554]: I think best way to test is 

- Mint sUSD in period 1
- Close period 1
- Should have rewards / fees claimable in period 1 
- Burn all debt 
- Should still have rewards / fees claimable in period 1

File: contracts/Issuer.sol [hunk 9]
```diff
@@ -523,6 +514,8 @@ contract Issuer is Owned, MixinSystemSettings, IIssuer {
     }
 
     function issueSynths(address from, uint amount) external onlySynthetix {
+        require(amount > 0, "Issuer: cannot issue 0 synths");
```
Comment (jacko125) [inline contracts/Issuer.sol:517]: Save them gas on this check

File: contracts/SynthetixDebtShare.sol [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+        emit Transfer(address(from), address(to), amount);
+    }
+
+    function importAddresses(address[] calldata accounts, uint256[] calldata amounts) external onlyOwner onlySetup {
+        for (uint i = 0; i < accounts.length; i++) {
+            mintShare(accounts[i], amounts[i]);
+        }
+    }
+
+    function finishSetup() external onlyOwner {
```
Comment (jacko125) [inline contracts/SynthetixDebtShare.sol:170]: Love how this is just a toggle for finalizing the import

**Review by jacko125 — APPROVED** [review]
LGTM most of the changes.

File: contracts/Issuer.sol (hunk 9 above)
Comment (dbeal-eth) [reply contracts/Issuer.sol:517]: if this check is not included, a test fails because the issuer tries to issue 0 synths and ultimately succeeds in doing so

File: contracts/SynthetixDebtShare.sol (hunk 10 above)
Comment (dbeal-eth) [reply contracts/SynthetixDebtShare.sol:170]: the owner calls it once it has finished calling `importAddresses`, at which point it will not be able to call `importAddresses` any more. You will see later this week, it will be included in the script.

its not flippable back to false intentionally, its just to record if the contract is still in a sort of initialization mode or not.

File: test/contracts/DebtCache.js [hunk 11]
```diff
@@ -582,8 +582,8 @@ contract('DebtCache', async accounts => {
 
 			it('When the debt snapshot is invalid, cannot issue, burn, exchange, claim, or transfer when holding debt.', async () => {
 				// Ensure the account has some synths to attempt to burn later.
-				await synthetix.transfer(account1, toUnit('1000'), { from: owner });
-				await synthetix.transfer(account2, toUnit('1000'), { from: owner });
+				await synthetix.transfer(account1, toUnit('10000'), { from: owner });
```
Comment (dbeal-eth) [inline test/contracts/DebtCache.js:585]: this test is changed because the amount of SNX sent initially is not enough to cover the case where issued synths are low enough to allow testing of the final assertion to not transfer SNX if issued debt

**Review by barrasso — APPROVED** [review]

**Review by artdgn — CHANGES_REQUESTED** [review]
Mainly small issues and questions.

File: contracts/SynthetixDebtShare.sol [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+        if (balance == 0) {
+            return 0;
+        }
+        
+        return balance.divideDecimal(totalSupplyOnPeriod[periodId]);
+    }
+
+    function allowance(address account, address spender) public view returns (uint) {
+        if (authorizedBrokers[spender]) {
+            return uint(-1);
```
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply contracts/SynthetixDebtShare.sol:106]: doesn't work in this version of solidity.

File: contracts/SynthetixDebtShare.sol [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+        if (accountPeriodHistoryCount == 0) {
+            return 0;
+        }
+
+        return balances[account][accountPeriodHistoryCount - 1].amount;
+    }
+
+    function balanceOfOnPeriod(address account, uint periodId) public view returns (uint) {
+        uint accountPeriodHistoryCount = balances[account].length;
+        for (int i = int(accountPeriodHistoryCount) - 1;i >= int(MAX_PERIOD_ITERATE < accountPeriodHistoryCount ? accountPeriodHistoryCount - MAX_PERIOD_ITERATE : 0);i--) {
```
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply contracts/SynthetixDebtShare.sol:71]: 1. an example? tbh I'm not seeing how this would be more readable
2. By the time, a caller would probably prefer that the function drop out in order to save gas rather than trying to binary search for really old values. Also binary search would still fail in the case of extremely large number of values (ex 60k). So rather than adding the functionality which still only works in some cases (which we don't really need), I just didn't add the bin search
3. the loop is required because it would not be gas efficient for fee pool to iterate from outside the contract
4. are you saying simplify the declaration by adding a second `if` condition within the loop? in that case, why not just add line before:
```
uint maxIterate = int(MAX_PERIOD_ITERATE < accountPeriodHistoryCount ? accountPeriodHistoryCount - MAX_PERIOD_ITERATE : 0);
```

File: contracts/SynthetixDebtShare.sol [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+    function addAuthorizedBroker(address authorizedBroker) external onlyOwner {
+        authorizedBrokers[authorizedBroker] = true;
+    }
+
+    function removeAuthorizedBroker(address authorizedBroker) external onlyOwner {
+        authorizedBrokers[authorizedBroker] = false;
+    }
+
+    function setCurrentPeriodId(uint newPeriodId) external onlyIssuer {
+        totalSupplyOnPeriod[newPeriodId] = totalSupplyOnPeriod[currentPeriodId];
```
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply contracts/SynthetixDebtShare.sol:124]: good idea

File: contracts/SynthetixDebtShare.sol [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+
+    function approve(address spender, uint256 amount) external {
+        revert("debt shares are not transferrable");
+    }
+
+    function transfer(address to, uint256 amount) external {
+        revert("debt shares are not transferrable");
+    }
+
+    function transferFrom(address from, address to, uint256 amount) external onlyAuthorizedBrokers {
```
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply contracts/SynthetixDebtShare.sol:157]: right now the only authorized broker we are going to have is the L2 bridge contract, for transferring debt to L2.

we can still add the functionality to let users "approve" a broker I suppose but it really seems like an unnecessary transaction

File: contracts/SynthetixDebtShare.sol [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+// Libraries
+import "./SafeDecimalMath.sol";
+
+// https://docs.synthetix.io/contracts/source/contracts/synthetixdebtshare
+contract SynthetixDebtShare is Owned, MixinResolver, ISynthetixDebtShare {
+    using SafeMath for uint;
+    using SafeDecimalMath for uint;
+
+    struct PeriodBalance {
+        uint amount;
```
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply contracts/SynthetixDebtShare.sol:17]: don't know how I missed this :laughing:

File: contracts/SynthetixDebtShare.sol (hunk 12 above)
[the remark this answers is not in the document]
Comment (artdgn) [reply contracts/SynthetixDebtShare.sol:106]: right right :+1:

File: contracts/SynthetixDebtShare.sol (hunk 13 above)
[the remark this answers is not in the document]
Comment (artdgn) [reply contracts/SynthetixDebtShare.sol:71]: I think if you don't want to add the binary search:

1. It would be nice to enable someone to perform the binary search externally by adding a length getter:
```solidity
function balancesLength(address account) external view returns (uint) {
    return balances[account].length;
}
```
I think given that this contract is not upgradeable, this would be an important addition to make the data easily usable from outside.

2. Just making the code block a bit more readable, and possibly adding a short circuit would be great. Maybe something like:
```solidity
        uint balancesLength = balances[account].length;
        // no history: short circuit everything
        if (balancesLength == 0) {
            return 0; 
        }
        PeriodBalance memory entry;
        uint maxIterate = balancesLength > MAX_PERIOD_ITERATE ? balancesLength - MAX_PERIOD_ITERATE : 0;
        for (uint i = balancesLength - 1; i >= maxIterate; i--) {
            entry = balances[account][i];
            if (entry.periodId <= periodId) {
                return entry.amount;
            }
        }
        return 0;
```

3. > Also binary search would still fail in the case of extremely large number of values (ex 60k).

Can be implemented with limits (what indices to search between), and 60k weeks is more than a 1000 years :)

File: contracts/SynthetixDebtShare.sol (hunk 15 above)
[the remark this answers is not in the document]
Comment (artdgn) [reply contracts/SynthetixDebtShare.sol:157]: To me it seems that the blast radius / surface area this `authorizedBroker` has, is very large in this contract and so the whole system. 

The reasons I think it makes sense to add the approve workflow:
1. It reduces the blast radius of any issue with `authorizedBroker` from "the whole protocol needs to stop and migrate" to "approvals need to be revoked and a few users need to be refunded"
2. It reduces the protocol's centralisation (in theory and practice).
3. Approvals transactions are cheap, and people are already used to them when bridging.
4. The workflow is right there in the standard that is being implemented, if the only reason to actively remove a security feature is minor gas savings - it may not be a good tradeoff.

File: contracts/SynthetixDebtShare.sol [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+
+    function balanceOfOnPeriod(address account, uint periodId) public view returns (uint) {
+        uint accountPeriodHistoryCount = balances[account].length;
+        for (int i = int(accountPeriodHistoryCount) - 1;i >= int(MAX_PERIOD_ITERATE < accountPeriodHistoryCount ? accountPeriodHistoryCount - MAX_PERIOD_ITERATE : 0);i--) {
+            if (balances[account][uint(i)].periodId <= periodId) {
+                return balances[account][uint(i)].amount;
+            }
+        }
+
+        return 0;
```
Comment (artdgn) [inline contracts/SynthetixDebtShare.sol:77]: Should this revert instead of returning 0? I know we don't like reverting views, a caller of this method has no way of knowing if users has actually 0 debt or if the loop went out of bounds.

An alternative is to return a tuple of `(uint amount, bool valid)` and to return `valid = false` if the loop went out of bounds.

File: contracts/SynthetixDebtShare.sol [hunk 18]
```diff
@@ -0,0 +1,225 @@
...
+        uint balance = balanceOf(account);
+
+        if (balance == 0) {
+            return 0;
+        }
+
+        return balance.divideDecimal(totalSupply());
+    }
+
+    function sharePercentOnPeriod(address account, uint periodId) external view returns (uint) {
```
Comment (artdgn) [inline contracts/SynthetixDebtShare.sol:94]: Same as `balanceOfOnPeriod` - this would return 0 if the loop goes out of bounds. I see same two solutions - either this reverts (because `balanceOfOnPeriod` reverts), or it returns another `bool valid` which the caller would use to know if the debt was really 0 or is just unknown.

File: contracts/SynthetixDebtShare.sol (hunk 15 above)
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply contracts/SynthetixDebtShare.sol:157]: you do realize that if a user has their debt share balance transferred from them, thats a good thing for them, right? the only attack vector we truly have to worry about here is that a user transfers their *own* debt to a place where they don't have to repay it, So really the only one who really should be approving is the protocol itself (hence `authorizedBroker`)

1) nobody needs to be refunded. at most, the protocol loosese money 2) debt shares are inherently completely centralized and controlled by us. if it was decentralized, the whole system wouldn't work 3) not cheap in friction and dev time :) 4) despite what it looks like, debt shares are not ERC20 compatible and they never will be, and if we treat it as such its just useless

File: contracts/SynthetixDebtShare.sol [hunk 19]
```diff
@@ -0,0 +1,225 @@
...
+
+    function finishSetup() external onlyOwner {
+        isInitialized = true;
+    }
+
+    /* ========== INTERNAL FUNCTIONS ======== */
+    function _supplyBalance(address account, uint amount) internal {
+        uint accountBalanceCount = balances[account].length;
+
+        if (accountBalanceCount == 0) {
```
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply contracts/SynthetixDebtShare.sol:178]: this code doesn't work correctly because, in the case that the `else` case is taken, the `lastBalance` is not taken into account from `prevPeriodId`

[the remark this answers is not in the document]
Comment (dbeal-eth) [reply contracts/SynthetixDebtShare.sol:178]: i understand that your concern is that this code is too hard to parse, so I will see if there is anything that can be done to reduce the comprehension barrier

File: contracts/SynthetixDebtShare.sol (hunk 16 above)
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply contracts/SynthetixDebtShare.sol]: 3783e05b

File: contracts/SynthetixDebtShare.sol [hunk 20]
```diff
@@ -0,0 +1,225 @@
...
+            mintShare(accounts[i], amounts[i]);
+        }
+    }
+
+    function finishSetup() external onlyOwner {
+        isInitialized = true;
+    }
+
+    /* ========== INTERNAL FUNCTIONS ======== */
+    function _supplyBalance(address account, uint amount) internal {
```
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply contracts/SynthetixDebtShare.sol]: 3783e05b

File: contracts/SynthetixDebtShare.sol [hunk 21]
```diff
@@ -0,0 +1,225 @@
...
+        uint accountBalanceCount = balances[account].length;
+
+        if (accountBalanceCount == 0) {
+            balances[account].push(PeriodBalance(amount, currentPeriodId));
+        }
+        else if (balances[account][accountBalanceCount - 1].periodId != currentPeriodId) {
+            balances[account].push(PeriodBalance(balances[account][accountBalanceCount - 1].amount.add(amount), currentPeriodId));
```
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply contracts/SynthetixDebtShare.sol]: didn't split into temp variables, but did split the line

this whole function will be re-evaluated to improve clarity in other ways.

3783e05b

File: contracts/SynthetixDebtShare.sol [hunk 22]
```diff
@@ -0,0 +1,225 @@
...
+
+    function transfer(address to, uint256 amount) external {
+        revert("debt shares are not transferrable");
+    }
+
+    function transferFrom(address from, address to, uint256 amount) external onlyAuthorizedBrokers {
+        _deductBalance(from, amount);
+        _supplyBalance(to, amount);
+
+        emit Transfer(address(from), address(to), amount);
```
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply contracts/SynthetixDebtShare.sol:162]: 3783e05b

File: test/contracts/SynthetixDebtShare.js [hunk 23]
```diff
@@ -0,0 +1,505 @@
...
+			await synthetixDebtShare.mintShare(account2, toUnit('20'), { from: issuer });
+
+			assert.bnEqual(await synthetixDebtShare.balanceOf(account1), toUnit('10'));
+			assert.bnEqual(await synthetixDebtShare.totalSupply(), toUnit('30'));
+		});
+
+		describe('on new period', async () => {
+			beforeEach(async () => {
+				await synthetixDebtShare.mintShare(account1, toUnit('10'), { from: issuer });
+				await synthetixDebtShare.setCurrentPeriodId(toUnit('10'), { from: issuer });
```
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply test/contracts/SynthetixDebtShare.js:126]: period can be any uint128

File: test/contracts/SynthetixDebtShare.js [hunk 24]
```diff
@@ -0,0 +1,505 @@
...
+	});
+
+	describe('setCurrentPeriodId()', () => {
+		it('fails when burning more than user has minted', async () => {
+			await onlyGivenAddressCanInvoke({
+				fnc: synthetixDebtShare.setCurrentPeriodId,
+				args: [toUnit('10')],
+				address: issuer,
+				accounts,
+				reason: 'SynthetixDebtShare: only issuer can mint/burn',
```
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply test/contracts/SynthetixDebtShare.js:219]: mislabel

[the remark this answers is not in the document]
Comment (dbeal-eth) [reply test/contracts/SynthetixDebtShare.js:217]: 6104a06b

File: contracts/SynthetixDebtShare.sol [hunk 25]
```diff
@@ -0,0 +1,225 @@
...
+        require(account != address(0), "ERC20: mint to the zero address");
+
+        _deductBalance(account, amount);
+
+        totalSupplyOnPeriod[currentPeriodId] = totalSupplyOnPeriod[currentPeriodId].sub(amount);
+        emit Transfer(account, address(0), amount);
+        emit Burn(account, amount);
+    }
+
+    function approve(address spender, uint256 amount) external {
```
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply contracts/SynthetixDebtShare.sol]: 6104a06b

File: contracts/SynthetixDebtShare.sol [hunk 26]
```diff
@@ -0,0 +1,225 @@
...
+
+    bytes32 public constant CONTRACT_NAME = "SynthetixDebtShare";
+
+    bytes32 private constant CONTRACT_ISSUER = "Issuer";
+
+    uint internal constant MAX_PERIOD_ITERATE = 10;
+
+    /* ========== STATE VARIABLES ========== */
+
+    mapping(address => bool) public authorizedBrokers;
```
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply contracts/SynthetixDebtShare.sol:32]: da67a273

File: contracts/SynthetixDebtShare.sol [hunk 27]
```diff
@@ -0,0 +1,225 @@
...
+
+    mapping(address => PeriodBalance[]) public balances;
```
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply contracts/SynthetixDebtShare.sol:40]: da67a273

File: contracts/SynthetixDebtShare.sol (hunk 17 above)
Comment (dbeal-eth) [reply contracts/SynthetixDebtShare.sol]: 1c8d00c

File: test/contracts/SynthetixDebtShare.js (hunk 23 above)
[the remark this answers is not in the document]
Comment (artdgn) [reply test/contracts/SynthetixDebtShare.js:116]: Yes, but it seems like a strange value and so it may be a sign of some kind of copy paste mix up when writing the test.

File: package.json [hunk 28]
```diff
@@ -77,14 +77,15 @@
 		"@eth-optimism/contracts": "0.3.4",
 		"@eth-optimism/core-utils": "~0.4.5",
 		"@eth-optimism/smock": "1.1.10",
-		"@gnosis.pm/safe-core-sdk": "1.3.0",
-		"@gnosis.pm/safe-service-client": "1.1.0",
```
Comment (barrasso) [inline package.json:81]: should we revert this and keep the upgraded gnosis sdk?

File: publish/deployed/mainnet/owner-actions.json [hunk 29]
```diff
@@ -1 +1,9 @@
-{}
```
Comment (barrasso) [inline publish/deployed/mainnet/owner-actions.json:1]: revert this?

File: publish/src/commands/owner.js [hunk 30]
```diff
@@ -396,9 +396,11 @@ const owner = async ({
 			]);
 			if (safeBatchSubmitter) {
 				await safeBatchSubmitter.appendTransaction({
-					to: OwnerRelayOnEthereum.address,
+					to: '0x0e16A6876210841577b233C4165d7B7EdF640b8a',
 					data: batchData,
 				});
+
+				console.log('go2');
```
Comment (barrasso) [inline publish/src/commands/owner.js:403]: a couple of console.logs in here

File: publish/src/commands/owner.js [hunk 31]
```diff
@@ -396,9 +396,11 @@ const owner = async ({
 			]);
 			if (safeBatchSubmitter) {
 				await safeBatchSubmitter.appendTransaction({
-					to: OwnerRelayOnEthereum.address,
+					to: '0x0e16A6876210841577b233C4165d7B7EdF640b8a',
```
Comment (barrasso) [inline publish/src/commands/owner.js:399]: unnecessary change here?

**Review by barrasso — COMMENTED** [review]
few minor comments
